# Programming In the Life Sciences

This is an OER around a six day course on using JavaScript to program the life sciences web, with a focus on the life sciences content in Wikidata and its API.

Description
-----------

In the life sciences the physical interactions between chemical entities, like genes, RNA, proteins, metabolites, and drugs, is of key interest. Not only do these interactions play an important role in the regulation of gene expression, inhibition of proteins, and they basically define all cellular processes. For example, pharmacology is the science studies the action of drugs on protein, and metabolism, similarly, depends on the interactions of small molecule substrates with enzymes.

With the increasing amount of knowledge and data in the life sciences, automation becomes increasingly important. Big data and small, complex data alike provide challenges to integrate data from different experiments and data sources. Various core life sciences databases provide SPARQL end point to their knowledge, while Wikidata is a spider in this web of semantic data. In this course, you will learn to use how to interact with at least SPARQL endpoints with JavaScript and visualize the results graphically with a library like d3.js or Cytoscape.js.

Schedule
--------

* Day 1: [Theory and HTML+JavaScript exercises](day1.md)
* Day 2: [Choosing a biological research question and exploring the Wikidata API](day2.md)
* Day 3: [Write down flows: data flow and programmatic (web service call) flow](day3.md)
* Day 4: [Final decision of visualisation type](day4.md)
* Day 5: [Implement application](day5.md)
* Day 6: [End presentations and individual assessment](day6.md)

Skills Description and Objectives
---------------------------------
In the life sciences the interactions between chemical entities is of key interest. Not only do these play an important role in the regulation of gene expression, and therefore all cellular processes, they are also one of the primary approaches in drug discovery. Pharmacology is the science studies the action of drugs, and for many common drugs, this is studying the interaction of small organic molecules and protein targets.

And with the increasing information in the life sciences, automation becomes increasingly important. Big data and small data alike, provide challenges to integrate data from different experiments. The Open PHACTS platform provides web services to support pharmacological research and in this course you will learn how to use such web services from programming languages, allowing you to link data from such knowledge bases to other platforms, such as those for data analysis.

The objectives of the course are:
* to have the ability to recognize various classes of chemical entities in pharmacology and to understand the basic physical and chemical interactions;
* to be familiar with technologies for web services in the life sciences;
* to obtain experience in using such web services with a programming language;
* to be able to select web services for a particular pharmacological question;
* to have sufficient background for further, more advanced, bioinformatics data analyses;
* to be familiar with modern software development practices.

During the practical you will develop a HTML+JavaScript application to answer a selected biological question. This question is determined as part of the practical, based on your own interest, with guidance from the instructor.

Literature
----------

* “Rang and Dale’s Pharmacology” by Rang et al. (Pearson, 7th edition, 2012).
* “JavaScript & jQuery: The Missing Manual” by D.S. McFarland (O'Reilly, 2nd edition, 2011)
* Codecademy's Learn HTML: [https://www.codecademy.com/learn/learn-html](https://www.codecademy.com/learn/learn-html)
* [Codecademy's Introduction to JavaScript](https://www.codecademy.com/learn/introduction-to-javascript)
* [OpenScienceMOOC's How to set up a repository on GitHub](https://github.com/OpenScienceMOOC/Module-5-Open-Research-Software-and-Open-Source/blob/master/content_development/Task_1.md)
* "Programming in the Life Sciences blog posts" by Willighagen: [http://chem-bla-ics.blogspot.com/search/label/%23pils](http://chem-bla-ics.blogspot.com/search/label/%23pils)
* [Wikidata Query Service Tutorial](https://wdqs-tutorial.toolforge.org/)
* [SPARQLing Biology: a beginners course](https://bigcat-um.github.io/SPARQLTutorialBioSB2019/) by Slenter (2019).
* List M, Ebert P, Albrecht F. [Ten Simple Rules for Developing Usable Software in Computational Biology](https://journals.plos.org/ploscompbiol/article?id=10.1371/journal.pcbi.1005265). Markel S, editor. PLOS Computational Biology. 2017 Jan 5;13(1):e1005265. 
* Perez-Riverol Y, Gatto L, Wang R, Sachsenberg T, Uszkoreit J, Leprevost F da V, et al. [Ten Simple Rules for Taking Advantage of Git and GitHub](https://journals.plos.org/ploscompbiol/article?id=10.1371/journal.pcbi.1004947). Markel S, editor. PLOS Computational Biology. 2016 Jul 14;12(7):e1004947. 
* Prlić A, Procter JB. [Ten Simple Rules for the Open Development of Scientific Software](https://journals.plos.org/ploscompbiol/article?id=10.1371/journal.pcbi.1002802). PLoS Computational Biology. 2012 Dec 6;8(12):e1002802.
* Sandve GK, Nekrutenko A, Taylor J, Hovig E. [Ten Simple Rules for Reproducible Computational Research](https://journals.plos.org/ploscompbiol/article?id=10.1371/journal.pcbi.1003285). Bourne PE, editor. PLoS Computational Biology. 2013 Oct 24;9(10):e1003285. 
* Wiegley, John. Git from the Bottum Up. 2018. [https://jwiegley.github.io/git-from-the-bottom-up/](https://jwiegley.github.io/git-from-the-bottom-up/)

